### PR TITLE
Add `--update` option to BaselineGenerator

### PR DESCRIPTION
**Pull request by dougbu**
- aspnet/AspNetCore-Internal#1649

### CODE REVIEW

File: eng/tools/BaselineGenerator/BaselineGenerator.csproj [hunk 1]
```diff
@@ -5,10 +5,11 @@
     <TargetFramework>netcoreapp2.1</TargetFramework>
     <StartArguments>-o "$(MSBuildThisFileDirectory)../../Baseline.Designer.props"</StartArguments>
     <StartWorkingDirectory>$(MSBuildThisFileDirectory)../../</StartWorkingDirectory>
+    <LangVersion>preview</LangVersion>
   </PropertyGroup>
 
   <ItemGroup>
-    <PackageReference Include="NuGet.Packaging" Version="4.8.0" />
+    <PackageReference Include="NuGet.Protocol" Version="4.8.2" />
```
Comment (dougbu) [inline eng/tools/BaselineGenerator/BaselineGenerator.csproj]: @Eilon please let me know if bumping the NuGet.Packaging version up a couple of patches or expanding the dependencies to include NuGet.Protocol is an issue.

File: eng/tools/BaselineGenerator/BaselineGenerator.csproj [hunk 2]
```diff
@@ -5,10 +5,11 @@
     <TargetFramework>netcoreapp2.1</TargetFramework>
     <StartArguments>-o "$(MSBuildThisFileDirectory)../../Baseline.Designer.props"</StartArguments>
     <StartWorkingDirectory>$(MSBuildThisFileDirectory)../../</StartWorkingDirectory>
+    <LangVersion>preview</LangVersion>
```
Comment (natemcmaster) [inline eng/tools/BaselineGenerator/BaselineGenerator.csproj]: > CSC : error CS1617: Invalid option 'preview' for /langversion. Use '/langversion:?' to list supported values. [D:\a\1\s\eng\tools\BaselineGenerator\BaselineGenerator.csproj]


Which version are you trying to use? I think the latest version supported by the 2.1 SDK is C# 7.3

Comment (dougbu) [reply eng/tools/BaselineGenerator/BaselineGenerator.csproj]: Hmm, I was actually building locally w/ the 2.2.203 SDK. I'll add a `global.json` and see if VS complains about my use of `using var`

Comment (dougbu) [reply eng/tools/BaselineGenerator/BaselineGenerator.csproj]: Either way, the Code Check step is failing w/ `CSC : error CS1617: Invalid option 'preview' for /langversion. Use '/langversion:?' to list supported values. [D:\a\1\s\eng\tools\BaselineGenerator\BaselineGenerator.csproj]`

Comment (natemcmaster) [reply eng/tools/BaselineGenerator/BaselineGenerator.csproj]: using var is a c# 8 feature, which our 2.x branches don't support. You will have to stick to c# 7

**Comment by dougbu** [comment]
:up::date: to get things compiling and working
- also added automatic detection of a V3 source

File: eng/tools/BaselineGenerator/Program.cs [hunk 3]
```diff
@@ -26,34 +32,57 @@ static void Main(string[] args)
...
+            _update = Option("-u|--update", "Regenerate the input (Baseline.xml) file.", CommandOptionType.NoValue);
 
             Invoke = () => Run().GetAwaiter().GetResult();
         }
 
         private async Task<int> Run()
         {
             var source = _source.HasValue()
-                ? _source.Value()
-                : "https://www.nuget.org/api/v2/package";
```
Comment (dougbu) [inline eng/tools/BaselineGenerator/Program.cs]: - This value wasn't an actual package source anyhow
- Should we use the V3 feed by default?

File: eng/tools/BaselineGenerator/BaselineGenerator.csproj (hunk 1 above)
Comment (Eilon) [reply eng/tools/BaselineGenerator/BaselineGenerator.csproj]: Not an issue, thanks for checking.

Comment (Eilon) [reply eng/tools/BaselineGenerator/BaselineGenerator.csproj]: (At least, I don't have an issue from a policy perspective. Not sure if there's any other kind of problem.)

File: eng/tools/BaselineGenerator/BaselineGenerator.csproj [hunk 4]
```diff
@@ -3,12 +3,11 @@
   <PropertyGroup>
     <OutputType>Exe</OutputType>
     <TargetFramework>netcoreapp2.1</TargetFramework>
-    <StartArguments>-o "$(MSBuildThisFileDirectory)../../Baseline.Designer.props"</StartArguments>
```
Comment (natemcmaster) [inline eng/tools/BaselineGenerator/BaselineGenerator.csproj]: Why is this removed?

File: eng/tools/BaselineGenerator/Program.cs [hunk 5]
```diff
@@ -26,34 +32,57 @@ static void Main(string[] args)
 
         private readonly CommandOption _source;
         private readonly CommandOption _output;
+        private readonly CommandOption _update;
 
         public Program()
         {
-            _source = Option("-s|--source <SOURCE>", "The NuGet v2 source of the package to fetch", CommandOptionType.SingleValue);
+            _source = Option("-s|--packageSource <SOURCE>", "The NuGet source of packages to fetch", CommandOptionType.SingleValue);
```
Comment (natemcmaster) [inline eng/tools/BaselineGenerator/Program.cs]:
```suggestion
            _source = Option("-s|--package-source <SOURCE>", "The NuGet source of packages to fetch", CommandOptionType.SingleValue);
```

Nit: follow POSIX conventions for capitalization of command line args

File: eng/tools/BaselineGenerator/README.md [hunk 6]
```diff
@@ -1,10 +1,20 @@
...
-=================
+# BaselineGenerator
 
 This tool is used to generate an MSBuild file which sets the "baseline" against which servicing updates are built.
 
 ## Usage
 
-1. Add to the [Baseline.xml](/eng/Baseline.xml) a list of package ID's and their latest released versions. The source of this information can typically
-  be found in the build.xml file generated during ProdCon builds. See https://github.com/dotnet/versions/blob/master/build-info/dotnet/product/cli/release/2.1.6/build.xml for example.
+Add `--packageSource {source}` to the commands below if the packages of interest are not all hosted on NuGet.org.
```
Comment (natemcmaster) [inline eng/tools/BaselineGenerator/README.md]:
```suggestion
Add `--package-source {source}` to the commands below if the packages of interest are not all hosted on NuGet.org.
```

File: eng/tools/BaselineGenerator/Program.cs (hunk 3 above)
Comment (natemcmaster) [reply eng/tools/BaselineGenerator/Program.cs]: > Should we use the V3 feed by default?

Sure, if it works.

**Review by natemcmaster — APPROVED** [review]
A few nit's and one question

File: eng/tools/BaselineGenerator/BaselineGenerator.csproj (hunk 4 above)
Comment (dougbu) [reply eng/tools/BaselineGenerator/BaselineGenerator.csproj]: 1. It conflicts with `--update`
2. It's not useful because `$(StartWorkingDirectory)` sets the current directory correctly and the tool defaults to finding `Baseline.Designer.props` in its current directory

File: eng/tools/BaselineGenerator/Program.cs [hunk 7]
```diff
@@ -26,34 +32,57 @@ static void Main(string[] args)
...
         {
             var source = _source.HasValue()
-                ? _source.Value()
-                : "https://www.nuget.org/api/v2/package";
-
-            var packageCache = Environment.GetEnvironmentVariable("NUGET_PACKAGES") != null
-                ? Environment.GetEnvironmentVariable("NUGET_PACKAGES")
-                : Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), ".nuget", "packages");
+                ? _source.Value().TrimEnd('/')
+                : "https://www.nuget.org/api/v2";
```
Comment (dougbu) [inline eng/tools/BaselineGenerator/Program.cs]:
```suggestion
                : "https://api.nuget.org/v3/index.json";
```

**Review by natemcmaster — APPROVED** [review]
